### PR TITLE
[api-minor] Add a new `transferPdfData` option to allow transferring more data to the worker-thread (bug 1809164)

### PR DESCRIPTION
**Pull request by Snuffleupagus**
Also, removes the `initialData`-parameter JSDocs for the `getDocument`-function given that this parameter has been completely unused since PR #8982 (over five years ago). Note that the `initialData`-parameter is, and always was, intended to be provided when initializing a `PDFDataRangeTransport`-instance.

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
This *seems* to work, but doesn't cover `initialData` yet.

/botio unittest

File: src/display/api.js [hunk 1]
```diff
@@ -375,6 +378,9 @@ function getDocument(src) {
         isValidFetchUrl(params.cMapUrl, document.baseURI) &&
         isValidFetchUrl(params.standardFontDataUrl, document.baseURI));
   }
+  if (typeof params.takeOwnershipOfData !== "boolean") {
+    params.takeOwnershipOfData = false;
```
Comment (calixteman) [inline src/display/api.js:382]: nit: Isn't enough to write: `params.takeOwnershipOfData = params.takeOwnershipOfData  === true;` ?

File: src/display/api.js [hunk 2]
```diff
@@ -513,6 +518,9 @@ async function _fetchDocument(worker, source, pdfDataRangeTransport, docId) {
     source.contentDispositionFilename =
       pdfDataRangeTransport.contentDispositionFilename;
   }
+  const transfers =
+    source.transferBinaryData && source.data ? [source.data.buffer] : null;
```
Comment (calixteman) [inline src/display/api.js:522]: Is there any reason to not transfer `initialData` ?

Comment (Snuffleupagus) [reply src/display/api.js:522]: It's already being transferred via the existing `PDFDataTransportStream` code, since that's really the only place where that option matters; see also the changes further down in this patch.

Comment (calixteman) [reply src/display/api.js:522]: Oh yes I missed that it wasn't used when calling `sendWithPromise`.

**Comment by Snuffleupagus** [comment]
Hopefully this should work correctly, and there's now basic unit-tests for all the relevant code-paths.

**Comment by calixteman** [comment]
What about: https://github.com/mozilla/pdf.js/blob/d6f63b5fb7cc63fa5d38ff71a56b8ee46d4f1b33/src/display/transport_stream.js#L66 ?

**Comment by Snuffleupagus** [comment]
> What about:
> https://github.com/mozilla/pdf.js/blob/d6f63b5fb7cc63fa5d38ff71a56b8ee46d4f1b33/src/display/transport_stream.js#L66

There's *somewhat similar* code found in all of the `src/display/fetch_stream.js`, `src/display/network.js`, and `src/display/node_stream.js` files, w.r.t. basically copying a TypedArray when reviving data, and it's not at all clear to me that it'd be generally safe to start changing that.

---

Even if we only change the mentioned `src/display/transport_stream.js` code, since that's probably the only thing that's generally safe, that code has quite limited test-coverage and I'm a bit worried about breaking the Firefox PDF Viewer in some weird edge-case.
*Ninja-edit:* Although looking the following code, I suppose that this *should* be safe in Firefox?
 - https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfStreamConverter.jsm#1042-1052
 - https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfStreamConverter.jsm#207-228

**Comment by calixteman** [comment]
In looking at:
https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfStreamConverter.jsm
I noticed that all the `postMessage` calls don't use the 3rd argument for transferable.
So right now, it means that we get a copy of the data in pdf.js side.
I do wonder if we should it: I think it should be safe (note that I don't have a strong certainty), wdyt ?

Anyway I'm fine to take the risk to break some corner cases: if something is broken it's an opportunity to improve and add some new tests so it isn't that bad :).

Even the initial goal of the bug I filed was mainly to fix a memleak, I think it's not that bad to improve that stuff here: it could help to win few 0.01% of battery on mobile.

File: src/display/transport_stream.js [hunk 3]
```diff
@@ -62,17 +76,20 @@ class PDFDataTransportStream {
     this._pdfDataRangeTransport.transportReady();
   }
 
-  _onReceiveData(args) {
-    const buffer = new Uint8Array(args.chunk).buffer;
-    if (args.begin === undefined) {
+  _onReceiveData({ begin, chunk }) {
```
Comment (calixteman) [inline src/display/transport_stream.js:79]: The chunk can be an `ArrayBuffer`: 
https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfJsNetwork.jsm#184
it's used in `onDone`:
https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfStreamConverter.jsm#717-725
and then:
https://github.com/mozilla/pdf.js/blob/d6f63b5fb7cc63fa5d38ff71a56b8ee46d4f1b33/src/display/transport_stream.js#L47

You can saw that with:
https://bibliotekacyfrowa.ujk.edu.pl//Content/4430/PDF/475183.pdf

We should really have consistent types everywhere...

Comment (Snuffleupagus) [reply src/display/transport_stream.js:79]: It seems that the code in `PdfStreamConverter.jsm` consistently, at least I think so, returns data in as a `Uint8Array`. Hence we might want to simply change the `PdfJsNetwork.jsm` code to be consistent with that?

Perhaps replacing the existing `getArrayBuffer` with something like:
```javascript
function getTypedArray(xhr) {
  const data = xhr.response;
  if (typeof data !== "string") {
    return new Uint8Array(data);
  }
  const length = data.length,
    array = new Uint8Array(length);
  for (let i = 0; i < length; i++) {
    array[i] = data.charCodeAt(i) & 0xff;
  }
  return array;
}
```
and update [this line](https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfJsNetwork.jsm#184) accordingly could work?

If so we should probably do that *first*, since the existing code in `src/display/transport_stream.js` should handle that (unless I'm missing something).

Comment (calixteman) [reply src/display/transport_stream.js:79]: I wrote the patch and it works with the pdf in bug 1809164.
Do you already have a wip you'll submit ? or should I submit mine ?

**Review by calixteman — APPROVED** [review]
LGTM. Thank you.

**Comment by Snuffleupagus** [comment]
> I noticed that all the `postMessage` calls don't use the 3rd argument for transferable.
> So right now, it means that we get a copy of the data in pdf.js side.
> I do wonder if we should it: I think it should be safe (note that I don't have a strong certainty), wdyt ?

For the data that's coming *directly* from the `PdfStreamConverter.jsm` file it really ought to be safe, given how that data is being read:
 - https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfStreamConverter.jsm#1042-1052
 - https://searchfox.org/mozilla-central/rev/a156a65ced2dae5913ae35a68e9445b8ee7ca457/toolkit/components/pdfjs/content/PdfStreamConverter.jsm#207-228

For the data that's coming from the `PdfJsNetwork.jsm` file, I suppose that it really depends on how the *browser* `XMLHttpRequest`-implementation works. Assuming that multiple `onreadystatechange` calls don't share one underlying ArrayBuffer, transferring this data ought to be safe as well.

Maybe we can just try updating the relevant `postMessage` calls with a `transfers`-argument, and test if that works?
However, we probably want to land *this PR* in mozilla-central first to not have all changes in the same Nightly-version.

**Comment by calixteman** [comment]
As far as I can tell, it'd be very strange to not have full ownership of the `ArrayBuffer` returned by xhr because it'd mean that the data could change if not read/copy immediately and in such a case I'd expected to have a big warning in the MDN doc.
And yes, we should change that stuff in 111 (it should be next week iirc).
